fem/fespace.hpp

```diff
@@ -330,6 +330,12 @@ class FiniteElementSpace
    virtual const Operator *GetProlongationMatrix() const
    { return GetConformingProlongation(); }
 
+   /// Return a prolongation operator that only acts on local dofs
+   /** The returned operator is owned by the FiniteElementSpace. This
+       differs from GetProlongationMatrix only in parallel. */
+   virtual const Operator *GetRestrictionTransposeOperator() const
+   { return GetConformingProlongation(); }
```

Comment:
Perhaps it would be more intuitive to provide `FiniteElementSpace::GetRestrictionOperator`, which returns an `Operator` that performs the exact same action as `FiniteElementSpace::GetRestrictionMatrix`, but potentially matrix-free. The caller can transpose using `TransposeOperator` (or `GetRestrictionTransposeOperator` can also be provided to avoid stacking transposes).

I think that would be easier to understand conceptually, and the fact that the transpose is implemented using `ConformingProlongationOperator` is (if I understand correctly) mostly done out of the convenience, and is an implementation detail that the user does not need to be aware of.

Reply:
I can certainly do this. Here's my hesitation.

`FiniteElementSpace` and `ParFiniteElementSpace` currently have `Operator * GetProlongationMatrix()` and `SparseMatrix* GetRestrictionMatrix()`. At some point in the transition to partial assembly, the return type of `GetProlongationMatrix()` was changed from `SparseMatrix*` to `Operator*`.

One option is what you propose, which results in having `Operator* GetRestrictionOperator()`, `SparseMatrix* GetRestrictionMatrix()`, and `Operator* GetProlongationMatrix()`, which seems confusing.

Another option is to change the name of `GetProlongationMatrix()` to `GetProlongationOperator()`, which is sensible but changes interfaces and potentially breaks user code (though I suspect this is almost exclusively used internally, I doubt users touch it much).

What I think is the "correct" thing to do (this is only becoming apparent to me now because of our discussions) is to change the return type of `GetRestrictionMatrix()` to `Operator*`. This will often be a `SparseMatrix`, and when it is not it could be implemented with `(Device)ConformingProlongationOperator` with a flag or with `TransposeOperator`. The problem here is that `GetRestrictionMatrix` is called ~40 times in the code base, more than a few of these calls seem to require an explicit assembled matrix, and I would need significant help and time to sort it out.

Reply:
I actually think that `Operator* GetRestrictionOperator()` and `SparseMatrix* GetRestrictionMatrix()` seem quite reasonable. The intent is conveyed pretty well by the name of each function.

What I agree is confusing is that `GetProlongationMatrix` may return something that isn't a matrix. That is definitely unfortunate, but it isn't related to this PR.

I would prefer to avoid a situation where there are two different, but similarly named functions that differ in some way that is not necessarily clear up-front (just as an example, `Mesh::GetNFaces` returns the 2D entities in a 3D mesh, `Mesh::GetNumFaces` returns the codim-1 entities in a general mesh — the user cannot infer this distinction from the difference between `N` and `Num`). Similarly, it might not be clear that the really important distinction between `GetRestrictionTransposeOperator` and `GetRestrictionMatrix` is not that one is transposed (which is really just an artifact of this particular implementation), but that one is matrix-free (the confusion is mitigated by the distinction between `Operator` and `Matrix`, but I think it's still present).

I also agree that the "best" interface would require breaking backwards compatibility (and so this is one instance where it would be useful to have some clear guidance on what the process is for introducing breaking API changes, cc: @tzanio), but I think there is still some room for improvement here without breaking changes.

Reply:
Just saw your comment pop up on my feed @barker29. I figured briefly mention that I'd probably be one of the few people where changing the name `GetProlongationMatrix` would break parts of my application code that deal with partial and element assembly methods. Although, I will say I only use that function like once, so it would be fairly simple fix.

Reply:
Thanks Will, that's very clear and you have convinced me. I will adopt your suggestion and we can deal with `Operator* GetProlongationMatrix` later.

Reply:
I believe this is resolved.